### ISSUE TITLE
[Bug] Yarn does not work with WSL / UNC paths

### ISSUE THREAD

**Issue by Stanzilla** [yarnpkg/berry#1230]
- [ ] I'd be willing to implement a fix

**Describe the bug**

Yarn does not work with WSL / UNC paths

**To Reproduce**

Run `yarn` on a project that is in a WSL container / UNC network share

**Screenshots**

![image](https://user-images.githubusercontent.com/75278/79890620-0e1efe00-8400-11ea-8427-422b238ee85c.png)

**Environment if relevant (please complete the following information):**

 - OS: Windows 10 2004
 - Node version 13.13.0
 - Yarn version [e.g. 2.0.0-rc1, ...] 2.0.0-rc.32

**Additional context**

**Comment by arcanis** [comment]
This is likely a matter of adding support for such paths to `fromPortablePath` / `toPortablePath`.

**Comment by Stanzilla** [comment]
@arcanis 

```
PS Microsoft.PowerShell.Core\FileSystem::\\wsl$\Ubuntu\home\stan\projects\personal\WeakAurasCompanion> yarn
Internal Error: EPERM: operation not permitted, mkdir '\\wsl$\Ubuntu\unc'
Error: EPERM: operation not permitted, mkdir '\\wsl$\Ubuntu\unc'
PS Microsoft.PowerShell.Core\FileSystem::\\wsl$\Ubuntu\home\stan\projects\personal\WeakAurasCompanion> 
```

**Comment by arcanis** [comment]
@Stanzilla We currently can't reproduce it - can you dig a bit and post the stacktrace for the problematic mkdir call?

**Comment by Stanzilla** [comment]
@arcanis where would I find the logs for those?

**Comment by Stanzilla** [comment]
@arcanis how would I do that?

**Comment by paul-soporan** [comment]
@Stanzilla Can you try the steps below and report back?
1. Create a new project with a package manager that works in WSL 2.
2. Add `@yarnpkg/fslib`, the package that contains our path and fs implementations.
3. Test the following things inside WSL 2:

```js
const {npath, xfs} = require(`@yarnpkg/fslib`);

// This is an alias for `C:\\` in Windows
const drivePath = npath.toPortablePath(`\\\\?\\C:\\`);
// Your WSL path
const wslPath = npath.toPortablePath(__dirname);

console.log(drivePath);
console.log(wslPath);

// Obviously, you don't have to give us the contents of your folders, just report if there's an error when executing these
console.log(xfs.readdirSync(drivePath));
console.log(xfs.readdirSync(wslPath));
```

The problem is that currently we can't reproduce your issue. Both the path implementation and the fs implementation seem to work.

**Comment by Stanzilla** [comment]
```
❯ node test.js  
\\?\C:\
/home/stan/projects/personal/WeakAurasCompanion
internal/fs/utils.js:230
    throw err;
    ^

Error: ENOENT: no such file or directory, scandir '\\?\C:\'
    at Object.readdirSync (fs.js:968:3)
    at NodeFS.readdirSync (/home/stan/projects/personal/WeakAurasCompanion/node_modules/@yarnpkg/fslib/lib/NodeFS.js:250:32)
    at Object.<anonymous> (/home/stan/projects/personal/WeakAurasCompanion/test.js:12:17)
    at Module._compile (internal/modules/cjs/loader.js:1123:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1143:10)
    at Module.load (internal/modules/cjs/loader.js:972:32)
    at Function.Module._load (internal/modules/cjs/loader.js:872:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at internal/main/run_main_module.js:17:47 {
  errno: -2,
  syscall: 'scandir',
  code: 'ENOENT',
  path: '\\\\?\\C:\\'
}
```

**Comment by paul-soporan** [comment]
My bad, I accidentally said "Run it in WSL 2" when I actually meant on Windows, but with the cwd inside the WSL container (Just like you originally tested Yarn when you opened the issue), so that you can test the UNC paths.

**Comment by Stanzilla** [comment]
```
PS Microsoft.PowerShell.Core\FileSystem::\\wsl$\Ubuntu\home\stan\projects\personal\WeakAurasCompanion> node index.js
internal/modules/cjs/loader.js:1023
  throw err;
  ^

Error: Cannot find module '\\wsl$\Ubuntu\home\stan\projects\personal\WeakAurasCompanion\index.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:1020:15)
    at Function.Module._load (internal/modules/cjs/loader.js:890:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at internal/main/run_main_module.js:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

**Comment by Stanzilla** [comment]
Turns out, this one was my bad, I did not run VSCode in WSL mode here, so that's why the paths were not working. Not sure if that warrants an extra warning/notice maybe? Or should yarn work outside of WSL mode?

**Comment by Stanzilla** [comment]
Tried again today:

```
➤ YN0000: Saving the new release in .yarn/releases/yarn-sources.js
➤ YN0001: Error: EPERM: operation not permitted, mkdir '\\wsl$\Ubuntu\unc'
➤ YN0000: Failed with errors in 2.91m
```

**Comment by merceyz** [comment]
I can't reproduce the issue after #1246 got merged so I'll close this.

---

> Tried again today:
> 
> ```
> ➤ YN0000: Saving the new release in .yarn/releases/yarn-sources.js
> ➤ YN0001: Error: EPERM: operation not permitted, mkdir '\\wsl$\Ubuntu\unc'
> ➤ YN0000: Failed with errors in 2.91m
> ```

@Stanzilla It looks like you're trying to update to the new version using the version that doesn't support WSL paths, if I open a WSL path and run `yarn set version berry && yarn set version 2.0.0-rc.36 && yarn` then it fails as expected since it doesn't contain the fix, now if I run `yarn set version 2.1.0` it also fails since `2.0.0-rc.36` is the one trying to write the update in the WSL path which it doesn't support. If you remove the `yarnPath` entry from `.yarnrc.yml` and run `yarn set version berry` it should work.
